### DISCUSSION TITLE
Support a pure GitOps workflow

### DISCUSSION THREAD

**Discussion by dabeeeenster** [Flagsmith/flagsmith#3983] [Ideas]
Some folk have asked if its possible to use Flagsmith in a pure GitOps workflow. What I take this to mean is that they want to store and manage Environment definitions 100% in code, and have that code then drive the flag engine. https://featurevisor.com/ and https://flagd.dev/ are examples of platforms that are designed around this workflow.

We already have our [Environment Document definition](https://docs.flagsmith.com/clients/overview#the-environment-document), which is currently represented as JSON.

Assuming people would not want to author an Environment Document 100% by hand, we could achieve this pretty easily if we had a way of converting a YAML/TOML based Environment Document into the JSON Environment Document that our clients can understand. This would be pretty easy to achieve using e.g. our Python or NodeJS SDK and providing some command line options to do the conversion.

Some thoughts:

- Would TOML be better than YAML? Or is there a better way to define this data? Can TOML encapsulate all of the current Environment Document JSON? 
- What about security/client side flags? We can use the [Edge Proxy](https://docs.flagsmith.com/advanced-use/edge-proxy) here. 
- Having YAML/TOML import and export at the Environment level would allow people to move to and from this workflow.
- We would need to have https://github.com/Flagsmith/flagsmith/issues/2024 implemented for full SDK support.

**Comment by matthewelwell** [comment]
@dabeeeenster I think this has some strong overlaps with our terraform provider. If you take the following statement in isolation then I would say that Terraform is the exact solution for this. 

> What I take this to mean is that they want to store and manage Environment definitions 100% in code

I guess there's an argument to say that Terraform adds an extra (perhaps unnecessary) layer onto the actual requirements when you also add the following additional clause. 

> and have that code then drive the flag engine

To extend this, one could for example, define the environment in terraform, then just add the following one line using our CLI into their CI/CD pipeline to achieve exactly what you're suggesting. 

```
FLAGSMITH_ENVIRONMENT=<key> flagsmith get -e environment
```

Then, literally the only thing that's a problem here is (some) people's disdain for TF since it went closed source. Is that enough of a reason, however, to reinvent a whole new solution here?

**Reply by dabeeeenster** [reply]
Hmm interesting. I guess the question then becomes one of what would a somewhat complex Environment definition look like in Terraform? Do we have an example of this?
